pkg/virt-controller/watch/pool/pool_test.go

```diff
@@ -686,10 +756,245 @@ var _ = Describe("Pool", func() {
+		})
+
+		It("should update VMs using opportunistic update strategy", func() {
+			pool, vm := DefaultPool(5)
+
+			pool.Spec.UpdateStrategy = &poolv1.VirtualMachinePoolUpdateStrategy{
+				Opportunistic: pointer.P(true),
+			}
+
+			oldPoolRevision := createPoolRevision(pool)
+
+			addPool(pool)
+			addCR(oldPoolRevision)
+
+			createVMsWithOrdinal(pool, 5, oldPoolRevision, oldPoolRevision, vm)
+
+			pool.Generation = 123
+			pool.Spec.VirtualMachineTemplate.Spec.Template.ObjectMeta.Labels = map[string]string{}
+			pool.Spec.VirtualMachineTemplate.Spec.Template.ObjectMeta.Labels["newkey"] = "newval"
+			newPoolRevision := createPoolRevision(pool)
+
+			updatePool(pool)
+
+			addCR(newPoolRevision)
+
+			expectUpdateVirtualMachines()
+
+			sanityExecute()
+
+			vms, err := fakeVirtClient.KubevirtV1().VirtualMachines(pool.Namespace).List(context.TODO(), metav1.ListOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(vms.Items).To(HaveLen(5))
+
+			for _, vm := range vms.Items {
+				Expect(vm.Spec.Template.ObjectMeta.Labels).To(HaveKeyWithValue("newkey", "newval"))
+			}
+
+			Expect(testing.FilterActions(&fakeVirtClient.Fake, "update", "virtualmachines")).To(HaveLen(5))
+		})
+
+		It("should update VMIs using node selector orderedpolicy proactive update strategy", func() {
+			pool, vm := DefaultPool(5)
+
+			pool.Spec.UpdateStrategy = &poolv1.VirtualMachinePoolUpdateStrategy{
+				Proactive: &poolv1.VirtualMachinePoolProactiveUpdateStrategy{
+					SelectionPolicy: &poolv1.VirtualMachinePoolSelectionPolicy{
+						OrderedPolicies: &poolv1.VirtualMachinePoolOrderedPolicy{
+							NodeSelectorRequirementMatcher: &[]k8sv1.NodeSelectorRequirement{
+								{
+									Key:      "node-role.kubernetes.io/worker",
+									Operator: k8sv1.NodeSelectorOpNotIn,
+									Values:   []string{"test"},
+								},
+							},
+						},
+					},
+				},
+			}
+
+			oldPoolRevision := createPoolRevision(pool)
+
+			pool.Generation = 123
+			pool.Spec.VirtualMachineTemplate.Spec.Template.ObjectMeta.Labels = map[string]string{}
+			pool.Spec.VirtualMachineTemplate.Spec.Template.ObjectMeta.Labels["newkey"] = "newval"
+			newPoolRevision := createPoolRevision(pool)
+
+			addPool(pool)
+
+			createVMsWithOrdinal(pool, 5, newPoolRevision, oldPoolRevision, vm)
+
+			addCR(oldPoolRevision)
+			addCR(newPoolRevision)
+			expectControllerRevisionCreation(newPoolRevision)
+
+			vms, err := controller.listVMsFromNamespace(pool.Namespace)
+			Expect(err).ToNot(HaveOccurred())
+
+			addNodeSelectorToVM(vms, pool)
+
+			sanityExecute()
+
+			pool, err = fakeVirtClient.PoolV1alpha1().VirtualMachinePools(pool.Namespace).Get(context.TODO(), pool.Name, metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(pool.Status.Replicas).To(Equal(int32(5)))
+
+			vmIList, err := fakeVirtClient.KubevirtV1().VirtualMachineInstances(pool.Namespace).List(context.TODO(), metav1.ListOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(vmIList.Items).To(HaveLen(2))
```

Comment:
Should updating not lead to 3 new VMs being scheduled?

Reply:
Just to confirm, in this case it should delete the three VMs that do not have the node selector and recreate them?

Reply:
Out of curiosity, is it a common use case to alter VMs manually that are also managed by a pool (i.e. adding labels)?